File: java/973.k-closest-points-to-origin.java
/*
 * @lc app=leetcode id=973 lang=java
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
class Solution {
    class PointDistance {
        int[] point;
        int distance;

        PointDistance(int[] point, int distance) {
            this.point = point;
            this.distance = distance;
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        int[][] res = new int[k][];

        PriorityQueue<PointDistance> pq = new PriorityQueue<PointDistance>(
            Comparator.comparingInt((PointDistance pd) -> pd.distance)
        );

        for(int[] p : points) {
            pq.add(new PointDistance(p, p[0] * p[0] + p[1] * p[1]));
        }

        for(int i = 0; i < k; i++) {
            res[i] = pq.poll().point;
        }

        return res;
    }
}
// @lc code=end

File: cpp/973.k-closest-points-to-origin.cpp
/*
 * @lc app=leetcode id=973 lang=cpp
 *
 * [973] K Closest Points to Origin
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        using T = tuple<int,int,int>;
        std::priority_queue<T, vector<T>, std::greater<T>> pq;
        
        for(vector<int>& point : points) {
            pq.emplace(point[0] * point[0] + point[1] * point[1]
                , point[0]
                , point[1]);
        }

        vector<vector<int>> res;

        while(k--) {
            res.push_back({std::get<1>(pq.top()), std::get<2>(pq.top())});
            pq.pop();
        }

        return res;
    }
};
// @lc code=end

